DEV-3611 - require-minimum-instances

piotrwalczak23 opened the pull request:
it will be used in jenkins script as follows:
```
    for STACK in ${DEPLOY_TARGETS}; do
    	if [[ $STACK == live-* ]]; then
           PREVIOUS_CAPACITY=$(devops/bin/exec stack require-minimum-instances ${STACK} 2)
        fi
        
        devops/bin/deploy codedeploy-push-and-wait ${STACK} devops/build/revisions/${STACK}

	if [[ $STACK == live-* ]]; then
           devops/bin/exec stack set-desired-capacity ${STACK} ${PREVIOUS_CAPACITY}
        fi
   done
```

dmitrygusev commented:
Looks like all you need from this change is to get number of desired instances, everything else can be done unconditionally right in the Jenkins script, so I'd rather implement `bin/stack get-desired-capacity ${STACK}` and added some logging around auto-scaling, as right now it's completely silent. We may also want to add timeout to the `wait-for-elb-instances`, otherwise the script will run forever.

```bash
for STACK in ${DEPLOY_TARGETS}; do
  local ORIGINAL_CAPACITY=$(devops/bin/exec stack get-desired-capacity ${STACK})

  if [[ $STACK == live-* ]]; then
    devops/bin/exec stack set-desired-capacity ${STACK} 2
    # this will print logs while waiting,
    # timeout after 600 seconds should fail the deployment
    devops/bin/exec stack wait-for-elb-instances ${STACK} 600
  fi
        
  devops/bin/deploy codedeploy-push-and-wait ${STACK} devops/build/revisions/${STACK}

  # this should also be fine to run unconditionally,
  # maybe omit the `if` just to not repeat the same condition twice?
  if [[ $STACK == live-* ]]; then
    devops/bin/exec stack set-desired-capacity ${STACK} ${ORIGINAL_CAPACITY}
  fi
done
```

Btw, there's a bug in scripts, namely in the `bin/deploy wait` and `bin/stack wait` that it doesn't print new line until there's a status change, it uses CR (`\r`) if status is the same -- that's a feature for terminal to print status update on the same line.

But when running on Jenkins, looks like that CR doesn't flush some buffers in Jenkins and you can't see current output in the Jenkins console until new line is printed, which is only after status change. So you can't really see what's going on with the script until it changes status. The output will be become visible if the script times out, but it would be nice to observe it in real time.

Maybe we could somehow detect if script is running in a terminal and supports CR, or we can just print new lines instead? Maybe something like this can fix it: https://rosettacode.org/wiki/Check_output_device_is_a_terminal#UNIX_Shell

piotrwalczak23 commented:
I have changed require-minimum-instances to get-desired-capacity and fixed new lines handling.

For timeout problem:
```timeout 600 devops/bin/exec stack wait-for-elb-instances ${STACK}```
will do.
